src/lib/web-rpde-query.js

```diff
@@ -0,0 +1,81 @@
+import Settings from './settings.js';
+import { database_pool } from './database.js';
+
+
+
+
+class rpde_query {
+  constructor(afterTimestamp, afterId, limit) {
+    this.afterTimestamp = parseInt(afterTimestamp) > 0 ? parseInt(afterTimestamp) : null;
+    this.afterId = afterId;
+    this.limit = parseInt(limit) > 0 ? parseInt(limit) : Settings.rpdeDefaultLimit;
+    this.nextAfterTimestamp = null;
+    this.nextAfterId = null;
+    this.rows = []
+    this.dontServeDataUntilSecondsOld = Settings.rpdeDontServeDataUntilSecondsOld;
+  }
+
+  async run() {
+    // ----- Build SQL
+    // Postgres will compare by microseconds if we let it. To make sure paging work correctly, we must compare by seconds only. Hence "extract(epoch" stuff.
+    let sql = "SELECT * FROM normalised_data ";
+    // This is to deal with the race condition described at https://developer.openactive.io/publishing-data/data-feeds/implementing-rpde-feeds#transactions-preventing-delayed-item-interleaving
+    sql += " WHERE updated_at < ((now() at time zone 'utc') - interval '"+ this.dontServeDataUntilSecondsOld +" seconds') ";
+    // Now if user has specified after, add those
+    let params = [];
+    if (this.afterTimestamp && this.afterId) {
+       sql += " AND ( ( extract(epoch from updated_at)::int = $1 AND data_id > $2) OR (extract(epoch from updated_at)::int > $1)) ";
+       params = [this.afterTimestamp, this.afterId];
+    }
+    // Finally order and limit
+    sql += "ORDER BY extract(epoch from updated_at)::int ASC, data_id ASC LIMIT " + this.limit
+    // ----- Run Query, store results
```

Comment:
is this.limit safe? it looks like it could come through the http get query parameters. Using the $ placeholders like you've done elsewhere looks like the right thing to do

Reply:
I guess it comes through parseInt and should at least be a number, though you could upset parseInt .

Reply:
Yes, because in the constructor it is def turned into a Int. But placeholder probably good for another safety thing.